requests-for-comments/2023-05-30-clickhouse-as-only-store.md

```diff
@@ -0,0 +1,35 @@
+# Request for comments: Clickhouse as our only data store
+
+## Problem statement
+
+- Having two separate places where we store essentially the same data (person/person_distinct_id/groups) means we often have inconsistencies
+- Our plans with the data warehouse are to allow customers to model all of their entities (persons/groups) in SQL in Clickhouse. For this to work we need to be able to read everything everywhere from clickhouse
```

Comment:
This analytics data is already in clickhouse right?

How would they access app data (non-analytics / their business logic) from the warehouse? I _assumed_ this would be a postgres connector, rather than loading this data into the warehouse?

Reply:
Not sure you saw my demo on monday, but the idea is people will be ETL'ing all sorts of data (Stripe, CRM, production database) into Clickhouse, and then using that to basically set person/group properties. This means PostHog becomes their source of truth, which would be incredibly powerful.

For the above to be even more powerful, we'd want to give users the option to use that data to serve feature flags and experiments. For that, Clickhouse needs to be the source of truth for person/groups data. Hence this RFC.

> Seems so far that the only real issue right now is postgres going down due to analytics load which brings down the entire app. Is this a fair summary of the problem?

That's one problem, yes, but the biggest opportunity is this modelling thing, which making Clickhouse our source of truth would enable.

Reply:
> unless you're imagining a flow where it's not just person/group properties, but any random ETL tables they generate inside the warehouse? In which case, I guess yeah figuring out a way to query clickhouse for flags becomes integral.

I am...

Reply:
One way (you're not going to like it though) might be to copy these random ETL props that are linked to persons into postgres.

We definitely can't go to clickhouse for every decide request: this is way too slow & also way too expensive for clickhouse.

Then comes caching: Either cache all person properties in Clickhouse, so access becomes cheap [but this needs a very large cache]; or cache flag responses for people [but this also needs a very large cache & isn't very useful because any random new anon user will miss the cache, and identified users are queried once anyway].

So caching doesn't seem like a viable option to me (unless we can get a Very Large Cache). Hence... copy all data to postgres & keep it there for fast flag access 🙈